0006-cdt-protocol/README.md

```diff
@@ -0,0 +1,205 @@
+
+CDT requirements:
+
+- Once a new endpoint configuration is pushed, there must be an EP configuration provider broadcasting an event over [NATS](http://nats.io/) about the new configuration availability.
+- The provider must receive a confirmation that the EP configuration was successfully pushed to the consumer.
+- When a consumer tries to pull an endpoint configuration that does not exist, or in case of other errors during configuration pull, HTTP status codes and arbitrary reason phrases must be used to inform about the errors occurred.
+
+## Design
+
+### Configuration pull
+
+Configuration pull is used when a service is intended to request particular configuration.
+
+No delivery confirmation is required for configuration pull, as endpoint can detect delivery fail and request configuration again.
+
+#### Subject structure
+
+The [EP configuration consumer](#language) should send messages using this NATS subject:
+```
+kaa.v1.service.{cdp-service-instance-name}.cmx2cdp.{message-type}
+```
+
+Also, the consumer should include NATS `replyTo` field pointing to the consumer service replica that will handle the response:
+```
+kaa.v1.replica.{cmx-service-instance-replica-id}.cmx2cdp.{message-type}
+```
+
+For more information, see [3/Messaging IPC][3/MIPC].
+
+#### Targeted message types
+
+There are two types of targeted messages:
+- `ConfigRequest` message is sent by EP configuration consumer.
+- `ConfigResponse` message is sent by [EP configuration provider](#language) service.
+
+`ConfigRequest` message structure:
+
+- `correlationId` (string, required): refer to [3/Messaging IPC][3/MIPC] RFC for description.<!--TODO-->
+- `timestamp` (number, required): message creation timestamp.
+- `timeout` (number, required): amount of time (starting from the timestamp) before the message gets ignored.
+- `endpointMessageId` (string, required): unique identifier of original endpoint message.
+- `appVersionName` (string, required): application version to which the endpoint configuration is applicable.
+- `endpointId` (string, required): unique identifier of endpoint to which the configuration is applicable.
+- `configId` (string, optional): unique identifier of endpoint configuration.
+If not present, response message will hold the latest version of configuration.
+
+Example:
+
+```json
+{
+  "correlationId": "07d78e95-2c4d-4899-957c-b9e5a3701fbb",
+  "timestamp": 1490303342158,
+  "timeout": 3000,
+  "endpointMessageId": "6b73cd7c-1de5-4c4e-bbad-b7eda079ccbd",
+  "appVersionName": "39774993-a426-4092-9e38-02ec213272d0",
+  "endpointId": "b197e391-1d13-403b-83f5-87bdd44888cf",
+  "configId": "76d34f8b-c038-413f-b122-318dce49edd1"
+}
+```
+
+`ConfigResponse` message structure:
+
+- `correlationId` (string, required): refer to [3/Messaging IPC][3/MIPC] RFC for description.<!--TODO-->
+- `timestamp` (number, required): message creation timestamp.
+- `timeout` (number, required): amount of time (starting from the timestamp) before the message gets ignored.
```

Comment:
In milliseconds, fixed. I think it's better to explain it here in a shorter wording than in 3/MIPC, for the sake of consistent UX. We already provided several links to 3/MIPC in this document, so the user should already know where to find the details.